consuming task results in finally

pritidesai opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

Final tasks can be configured to consume `results` of `PipelineTasks` from `tasks` section.

Based on the TEP, a final task failing to resolve task results is added to the list of `skippedTasks` and warning is recorded in controller logs.

```
unable to find result referenced by param "param1" in "final-task" ...
```

Implements TEP [#0004](https://github.com/tektoncd/community/blob/master/teps/0004-task-results-in-final-tasks.md)
Closes #2557 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```release-note
Final tasks can be configured to consume `results` of `PipelineTasks` from `tasks` section. 
```

pritidesai commented:
/kind feature

jerop commented on docs/pipelines.md:
```diff
@@ -688,6 +688,36 @@ spec:
+        - name: before-count
+          value: $(tasks.count-comments-before.results.count) #invalid
+        - name: after-count
+          value: $(tasks.count-comments-after.results.count) #invalid
```
i think the `#invalid` comments should be removed

pritidesai replied:
great catch @jerop, done 👍

jerop commented on pkg/reconciler/pipelinerun/resources/pipelinerunstate.go:
```diff
@@ -305,6 +311,18 @@ func (state PipelineRunState) GetTaskRunsStatus(pr *v1beta1.PipelineRun) map[str
 				})
 			}
 		}
+		// Maintain a TaskRun Object in pr.Status for a finally task which could not resolve task results
```
saw that this was already agreed upon in the TEP, but I think it might be worthwhile to explore adding these final tasks to skipped tasks (or another separate section for final tasks with invalid results) -- instead of creating a TaskRun object for a final task that wasn't executed 

related discussion: https://github.com/tektoncd/pipeline/issues/3139

pritidesai replied:
yes thats an alternative, another alternative could be: Tekton controller replacing such unresolved param with type default which is empty string for string type params. Let that final task execute expecting that the task author has implemented sanitization on such input params. @bobcatfish thoughts?

jerop reviewed (commented):
excited to see results supported in finally tasks! 

we can look into extending it to support guarding execution of final tasks based on task results using when expressions

vdemeester reviewed (commented):
/lgtm

jerop commented on pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:
```diff
@@ -69,6 +69,9 @@ type ResolvedPipelineRunTask struct {
 	ResolvedTaskResources *resources.ResolvedTaskResources
 	// ConditionChecks ~~TaskRuns but for evaling conditions
 	ResolvedConditionChecks TaskConditionCheckState // Could also be a TaskRun or maybe just a Pod?
+	// this flag is set for a task when task result resolution fails for that task
+	// and mainly used to create a task run object with failure for that task
+	InvalidTaskResultsInFinally bool
```
wondering if there's a way to avoid saving execution state in resolvedpipelineruntask because it no longer means that it's a PipelineTask with its references resolved (looking into refactoring so that we don't put the taskrun in this struct as well)....we'd initially considered saving the skipped state here when working on when expressions, but decided with calling the method twice (when skipping and when updating status)

cc @bobcatfish

pritidesai replied:
PipelineTask being skipped due to when expression failure is a different scenario then final task skip/failing due to task result resolution failure.

When expression is part of `ResolvedPipelineRunTask` and can be evaluated from `GetPipelineConditionStatus` i.e. we can call `Skip` again and will receive accurate result (once a PipelineTask is declared skipped, `Skip` would return true for that PipelineTask based on WhenExpression evaluation inside of `Skip`).

With final task being skipped/failed due to task results resolution failure is determined while retrieving execution queue and not scheduled if task result resolution fails. There is no means of storing such information in `ResolvedPipelineRunTask` except this flag `InvalidTaskResultsInFinally`. 

The key difference is a source of truth available in `ResolvedPipelineRunTask` to all the reconciling cycles v/s reconciler making decision in one cycle and making it available all subsequent cycles.

I hope this make sense 😜 

I will be happy to hear if we can still avoid such flags (not in favor of it either).

jerop replied:
Yes, that makes sense

Maybe in a future PR after we do some refactoring, we can avoid having that flag there

anderoo commented:
Happy to see this PR. Moves us closer to being able to conditionally execute finally tasks.

pritidesai commented:
Please review whenever you can @tektoncd/core-maintainers, would like to get this in 0.17 🙏

sbwsg commented:
Small suggestion, doesn't have to be a blocker on this PR, but it would be great to see an example yaml added to examples/v1beta1/pipelineruns for this too if possible.

/approve

pritidesai commented:
thanks @sbwsg, thats a great suggestion. Putting it on a brief hold until I add and end-to-end example.
/hold

pritidesai commented:
/hold cancel

@sbwsg @jerop I have updated this PR with following changes:

1. Include an [example](https://github.com/tektoncd/pipeline/pull/3242/files#diff-56c052a6b622ca906e10799a94023df4R167) with finally task utilizing result.

2. Updated the [flag](https://github.com/tektoncd/pipeline/pull/3242/files#diff-4e301064fef70f3c0db810267c0959f2R75) `InvalidTaskResultsInFinally` from `bool` to `string` to include the error message in the taskRun object in pipelineRun Status instead of just getting added to controller logs.

3. Populate `pipelineRunState` with such [taskRun](https://github.com/tektoncd/pipeline/pull/3242/files#diff-4e301064fef70f3c0db810267c0959f2R367) object from pipelineRun Status along with [the unit test](https://github.com/tektoncd/pipeline/pull/3242/files#diff-4514daff4848977b060192ad3a3694e4R1855). Discovered an issue with this PR while adding end-to-end example 😢 . Without these changes, `GetPipelineRunConditionStatus` iterating over `pipelineRunState` has no knowledge of a taskRun being declared as failure when such taskRun does not exist in `state`.

sbwsg commented:
Awesome, thanks Priti, looks good to me!

popcor255 commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -3925,6 +3925,132 @@ func TestReconcilePipeline_TaskSpecMetadata(t *testing.T) {
 	}
 }
 
+func TestReconcileWithTaskResultsInFinalTasks(t *testing.T) {
```
I believe we are trying to deprecate test builders from our tests. :sweat_smile: https://github.com/tektoncd/pipeline/issues/3178

popcor255 commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -3925,6 +3925,132 @@ func TestReconcilePipeline_TaskSpecMetadata(t *testing.T) {
+				tb.TaskRunTaskRef("hello-world"),
+				tb.TaskRunServiceAccountName("test-sa"),
+			),
+			tb.TaskRunStatus(
+				tb.StatusCondition(
+					apis.Condition{
+						Type:   apis.ConditionSucceeded,
+						Status: corev1.ConditionFalse,
+					},
+				),
+			),
+		),
+	}
+	d := test.Data{
+		PipelineRuns: prs,
+		Pipelines:    ps,
+		Tasks:        ts,
+		TaskRuns:     trs,
+	}
+	prt := NewPipelineRunTest(d, t)
+	defer prt.Cancel()
+
+	reconciledRun, clients := prt.reconcileRun("foo", "test-pipeline-run-final-task-results", []string{}, false)
+
+	expectedTaskRunName := "test-pipeline-run-final-task-results-final-task-1-9l9zj"
+	expectedTaskRun := tb.TaskRun(expectedTaskRunName,
+		tb.TaskRunNamespace("foo"),
+		tb.TaskRunOwnerReference("PipelineRun", "test-pipeline-run-final-task-results",
+			tb.OwnerReferenceAPIVersion("tekton.dev/v1beta1"),
+			tb.Controller, tb.BlockOwnerDeletion,
+		),
+		tb.TaskRunLabel("tekton.dev/pipeline", "test-pipeline"),
+		tb.TaskRunLabel("tekton.dev/pipelineRun", "test-pipeline-run-final-task-results"),
+		tb.TaskRunLabel("tekton.dev/pipelineTask", "final-task-1"),
+		tb.TaskRunSpec(
+			tb.TaskRunTaskRef("final-task-1"),
+			tb.TaskRunServiceAccountName("test-sa-0"),
+			tb.TaskRunParam("finalParam", "aResultValue"),
+		),
+	)
```
I believe we are trying to deprecate test builders from our tests. :sweat_smile: #3178 
It's not a blocker. I would like to lgtm this. I am glad to make this change later, if you want to merge this now.

popcor255 commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -3925,6 +3925,132 @@ func TestReconcilePipeline_TaskSpecMetadata(t *testing.T) {
 
+func TestReconcileWithTaskResultsInFinalTasks(t *testing.T) {
+	names.TestingSeed()
+	ps := []*v1beta1.Pipeline{tb.Pipeline("test-pipeline", tb.PipelineNamespace("foo"), tb.PipelineSpec(
+		tb.PipelineTask("dag-task-1", "dag-task"),
+		tb.PipelineTask("dag-task-2", "dag-task"),
+		tb.FinalPipelineTask("final-task-1", "final-task-1",
+			tb.PipelineTaskParam("finalParam", "$(tasks.dag-task-1.results.aResult)"),
+		),
+		tb.FinalPipelineTask("final-task-2", "final-task-2",
+			tb.PipelineTaskParam("finalParam", "$(tasks.dag-task-2.results.aResult)"),
+		),
+	))}
```
I believe we are trying to deprecate test builders from our tests. :sweat_smile: #3178 
It's not a blocker. I would like to lgtm this. I am glad to make this change later, if you want to merge this now.

pritidesai replied:
thanks @popcor255 yes, will change that later along with the other tests.

bobcatfish replied:
pinging this again

pritidesai replied:
🤦‍♀️  shying away in this PR

vdemeester reviewed (commented):
/lgtm

vdemeester commented:
```
2020-10-07T09:32:59.866Z	DEBUG	TestReconcileTaskResourceResolutionAndValidation/Fail_ValidateResolvedTaskResources	taskrun/reconciler.go:322	Updating status with:   v1beta1.TaskRunStatus{
  	Status: v1beta1.Status{
  		ObservedGeneration: 0,
- 		Conditions:         nil,
+ 		Conditions: v1beta1.Conditions{
+ 			{
+ 				Type:               "Succeeded",
+ 				Status:             "False",
+ 				LastTransitionTime: apis.VolatileTime{Inner: s"2020-10-07 09:32:59.864401656 +0"...},
+ 				Reason:             "TaskRunValidationFailed",
+ 				Message:            "invalid input resources for task test-task-missing-resource: Tas"...,
+ 			},
+ 		},
  		Annotations: nil,
  	},
  	TaskRunStatusFields: v1beta1.TaskRunStatusFields{
  		PodName:        "",
- 		StartTime:      nil,
+ 		StartTime:      s"2020-10-07 09:32:59.863482481 +0000 UTC m=+23.220666862",
  		CompletionTime: nil,
  		Steps:          nil,
  		... // 3 identical fields
  		TaskRunResults: nil,
  		Sidecars:       nil,
- 		TaskSpec:       nil,
+ 		TaskSpec:       &v1beta1.TaskSpec{Resources: &v1beta1.TaskResources{Inputs: []v1beta1.TaskResource{{...}}}},
  	},
  }

2020-10-07T09:32:59.879Z	DEBUG	TestReconcileTaskResourceResolutionAndValidation/Fail_ValidateResolvedTaskResources	taskrun/reconciler.go:322	Updating status with:   v1beta1.TaskRunStatus{
  	Status: v1beta1.Status{
  		ObservedGeneration: 0,
- 		Conditions:         nil,
+ 		Conditions: v1beta1.Conditions{
+ 			{
+ 				Type:               "Succeeded",
+ 				Status:             "False",
+ 				LastTransitionTime: apis.VolatileTime{Inner: s"2020-10-07 09:32:59.864401656 +0"...},
+ 				Reason:             "TaskRunValidationFailed",
+ 				Message:            "invalid input resources for task test-task-missing-resource: Tas"...,
+ 			},
+ 		},
  		Annotations: nil,
  	},
  	TaskRunStatusFields: v1beta1.TaskRunStatusFields{
  		PodName:        "",
- 		StartTime:      nil,
+ 		StartTime:      s"2020-10-07 09:32:59.863482481 +0000 UTC m=+23.220666862",
  		CompletionTime: nil,
  		Steps:          nil,
  		... // 3 identical fields
  		TaskRunResults: nil,
  		Sidecars:       nil,
- 		TaskSpec:       nil,
+ 		TaskSpec:       &v1beta1.TaskSpec{Resources: &v1beta1.TaskResources{Inputs: []v1beta1.TaskResource{{...}}}},
  	},
  }
```
Some unit errors *seem* legit ? :thinking: 
/lgtm cancel

pritidesai commented:
```
fatal: unable to access 'https://github.com/tektoncd/pipeline.git/': Could not resolve host: github.com
```
/retest

pritidesai commented:
> Some unit errors seem legit ? 🤔
/lgtm cancel

😭  Yup, recent commits caused the failures, fixed now 🙏

popcor255 commented:
/lgtm Thank you so much for your patience! :pray:

popcor255 commented:
/lgtm

popcor255 commented:
:cry: I check the issues to see if this is a flake. This looks legit error. Im not sure why? :confused: /lgtm cancel

popcor255 commented:
/lgtm cancel

pritidesai commented:
@popcor255 I have been seeing this kind of intermittent failure quite recently on other [PRs](https://github.com/tektoncd/pipeline/pull/3353#issuecomment-705080791) as well. 

```
Expected 2 number of successful events from pipelinerun and taskrun but got 3; list of received events
```

/test pull-tekton-pipeline-integration-tests

pritidesai commented:
/test pull-tekton-pipeline-integration-tests

pritidesai commented:
@popcor255 tests are all green now 💃  PTAL

vdemeester commented:
/lgtm

afrittoli commented on docs/pipelines.md:
````diff
@@ -699,6 +699,36 @@ spec:
+        - name: after-count
+          value: $(tasks.count-comments-after.results.count)
+```
+**Note** The scheduling of such final task does not change, it will still be executed in parallel with other
````
NIT: Missing ':' after **Note**

afrittoli commented on docs/pipelines.md:
```diff
@@ -699,6 +699,36 @@ spec:
 
+#### Consuming `Task` execution results in `finally`
+
+Final tasks can be configured to consume `Results` of `PipelineTask` from `tasks` section:
```
NIT: from *the* `tasks` section

afrittoli commented:
Thanks you for this! I really look forward to using this in our `dogfooding` cluster.

I'm a bit concerned about the final task failure on missing results. 
It matches what specified in the TEP, so unfortunately I'm a bit late to the discussion:
```
Q. What happens to Pipeline when the dependent task boskos-acquire either failed or not executed and the task result leased-resource is not initialized?

A. The finally Task boskos-release is attempted but declared failure since the task result leased-resource is not initialized. Pipeline controller logs the reason of this failure as invalid result reference InvalidTaskResultReference. Pipeline continues executing rest of the final tasks and exits with failure.
```

I have two concerns:
- From a pipeline run status point of view, we're now hosting in the pipeline run status a reference to a taskrun that was never created (at least that's my understanding). We're already doing that for `Conditions` and it's really confusing and hard to parse for both humans and UIs (dashboard / tkn). I'd hope we would do less of that and not more.
- From a workflow pov, I'd rather skip a finally task that cannot be executed because of missing results. The outcome of the pipeline will then match the reason for the results to be missing:
  - if the results are missing because the producer task failed, the pipeline will fail
  - if the results are missing because the producer was skipped:
     - if the producer was skipped without failures, the pipeline may still pass
     - if the producer was skipped because of some failure, the pipeline will fail

If we allowed using `when` expressions on finally tasks, a way to avoid failing a pipeline because of missing results could be to specify all required when expressions on the task, however that may not be trivial:
```
[Task A](guarded by X) -> [Task B] -> [Task C](produce result Z), [Finally Task D](consumes result Z)
```

With the current implementation, if A is skipped, B and C will not run, D will fail and the whole pipeline will fail.
The alternative for the pipeline author would be to guard D with X, but that does not feel to me like a good user experience.

If there is general agreement on the current approach, I'm ok with it.
In any case we should clearly document the behaviour in our docs, so that pipeline authors may know what to expect.

afrittoli commented on pkg/reconciler/pipelinerun/pipelinerun.go:
```diff
@@ -521,7 +521,20 @@ func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.Pip
+	// Before creating TaskRun for scheduled final task, check if it's consuming a task result
+	// Resolve and apply task result wherever applicable, report failure in case resolution fails
+	for _, rprt := range finallyRprts {
+		resolvedResultRefs, err := resources.ResolveResultRefs(pipelineRunFacts.State, resources.PipelineRunState{rprt})
+		if err != nil {
+			rprt.InvalidTaskResultsInFinally = err.Error()
+			logger.Warnf("Declaring final task %q failed as it failed to resolve task params for %q with error %v", rprt.PipelineTask.Name, pr.Name, err)
+			continue
+		}
+		resources.ApplyTaskResults(resources.PipelineRunState{rprt}, resolvedResultRefs)
```
It feels a bit un-natural to use two functions that work on slices, passing them elements of a slice one by one embedded in 1-element slices.
I wonder if it would make sense to make a version of those functions that work on single elements, and use that both here as well as for the implementation of the current functions that work on slices.

afrittoli replied:
Not something that needs to be fixed in the PR, it's also fine to change it as a follow-up if you agree.

pritidesai replied:
I have introduced a new function to handle single task for now, `taskResults` is due for refactoring. After @sbwsg's pipelineRun results PR #3472 is merged, I will take up refactoring `taskResults`. Changes in PR #3472 have simplified resolving pipelineRun results and cleaned up `resultrefresolution.go`.

vdemeester commented:
/lgtm cancel
/hold

pritidesai commented:
> From a pipeline run status point of view, we're now hosting in the pipeline run status a reference to a taskrun that was never created (at least that's my understanding). We're already doing that for Conditions and it's really confusing and hard to parse for both humans and UIs (dashboard / tkn). I'd hope we would do less of that and not more.

Without creating such object in `PipelineRun` status, how could we inform the pipeline author what happened to that final task without accessing controller logs? Why that final was not executed?

bobcatfish commented:
Thanks for raising this @afrittoli - I had agreed with the initial proposal but I'm not sure I fully realized this meant there would be a failed TaskRun created as well and I also hope we can find some different way of doing this.

> Without creating such object in PipelineRun status, how could we inform the pipeline author what happened to that final task without accessing controller logs? Why that final was not executed?

Some ideas: 
1. In slack @jerop mentioned maybe considering the Task "skipped" and including it in "skipped tasks".
2. Another possibility might be to have some new section specifically for finally tasks that don't execute.
3. Another possible approach would be go the opposite direction from what Andrea is suggested, e.g. embrace having a 'placeholder' taskrun in the status, e.g. when a pipelinerun starts, start building up 'placeholder' taskruns in the status for _all_ task
4. We could revisit the shape of the PipelineRun status entirely and how TaskRuns are represented - imo we haven't paid as much attention to status design as we have to spec and this has all kind of built up ad-hoc

I'm kind of inclined toward (4) (pre-GA), with maybe something like (1) in the meantime to unblock this work?

pritidesai commented:
> I'm kind of inclined toward (4) (pre-GA), with maybe something like (1) in the meantime to unblock this work?

`finally` task must execute and can not be skipped based on the original contract. The best way to handle this would be continue execution as is with default type value i.e. empty string in case of a string type and empty array in case of array type. Leave it up to pipeline author to decide how he wants to deal with such zero values. Make pipeline author aware of this kind of functionality by documenting it.

Going with (1) in meanwhile but we will get into backward compatibility cycle. And lots of use cases are eyeing on this feature so would like to get in ASAP.

bobcatfish commented:
> finally task must execute and can not be skipped based on the original contract.

Apologies for the frustration but I think it might be worth revisiting this contract :(

> The best way to handle this would be continue execution as is with default type value i.e. empty string in case of a string type and empty array in case of array type. Leave it up to pipeline author to decide how he wants to deal with such zero values. Make pipeline author aware of this kind of functionality by documenting it.

We discussed this option a fair amount iirc - doing this would mean changing a different "contract" i.e. what it means to express that a Task needs the result of another Task: at the moment the contract is that if the Task runs, it will be given that result from the previously executed Task.

Either way we are going to be changing a contract.

Maybe we can discuss more in the API working group on monday?

bobcatfish commented:
>  We're already doing that for Conditions and it's really confusing and hard to parse for both humans and UIs (dashboard / tkn).

one counter point @afrittoli : if the UI did show this finally task as "failed" when what happened was it really 'failed to execute b/c the result wasn't available' that's not quite as bad as what we see with conditions b/c in this case, we already have a failed pipelinerun. the conditions case was especially confusing b/c the condition task "failed" but the pipelinerun did not fail

pritidesai commented:
> Apologies for the frustration but I think it might be worth revisiting this contract :(

@bobcatfish Please don't apologize, no frustration at all, its part of the development cycle 😉  Its better to find the solution with everyone's agreement rather than building something which no one would like to use 😆 

> Either way we are going to be changing a contract.

🤣  🤣  🤣  

Yup, API WG it is.

pritidesai commented:
The way its implemented in this PR, a sample pipelinerun with two final tasks, `final-task-1` and `final-task-2` where `final-task-2` has a reference to task result of a `dag` task and that `dag` task failed:

```
tkn pr describe
Name:              write-and-cleanup-workspace
Namespace:         default
Pipeline Ref:      clone-cleanup-workspace
Service Account:   default
Timeout:           1h0m0s
Labels:
 tekton.dev/pipeline=clone-cleanup-workspace

🌡️  Status

STARTED         DURATION     STATUS
6 minutes ago   17 seconds   Failed

💌 Message

Tasks Completed: 3 (Failed: 2, Cancelled 0), Skipped: 0 (Failed to resolve task params for finally Task final-task-2 in PipelineRun write-and-cleanup-workspace: unable to find result referenced by param "commit" in "final-task-2": could not find successful taskrun for task "clone-app-repo")

📦 Resources

 No resources

⚓ Params

 No params

📝 Results

 No results

📂 Workspaces

 NAME           SUB PATH   WORKSPACE BINDING
 ∙ git-source   ---        VolumeClaimTemplate

🗂  Taskruns

 NAME                                                 TASK NAME        STARTED         DURATION     STATUS
 ∙ write-and-cleanup-workspace-final-task-2-fswll     final-task-2     ---             ---          Failed(InvalidTaskResultsInFinallyTask)
 ∙ write-and-cleanup-workspace-final-task-1-79mlt     final-task-1     6 minutes ago   11 seconds   Succeeded
 ∙ write-and-cleanup-workspace-clone-app-repo-qp2w5   clone-app-repo   6 minutes ago   5 seconds    Failed
```

<img width="1185" alt="Screen Shot 2020-10-08 at 1 46 05 PM" src="https://user-images.githubusercontent.com/206285/95512351-8f962880-096d-11eb-83db-475c6c225f24.png">

pritidesai commented:
@bobcatfish @afrittoli @jerop @sbwsg 

As I am thinking more and more about the use cases we are trying support with this feature, I am not able to think beyond `cleanup` (the one and only use case specified in the [TEP](https://github.com/tektoncd/community/blob/master/teps/0004-task-results-in-final-tasks.md#motivation)):

UC1:

The resources are acquired by the setup `Task` which has the names of resources. The clean up `Tasks`
must clean up only those resources. Also, cleanup must be done only if the setup was successful but regardless of the failure of the any other `Tasks` in the `Pipeline`.

```none
setup
  |
  | e.g. Project Name
  v
cleanup
```

Similar [use case](https://github.com/tektoncd/pipeline/issues/2557#issuecomment-692078619) from @spikeburton.

For this use case, if resource acquisition fail, ideally cleanup should be skipped. Cleanup can not be declared failure since the cleanup itself didn't fail. To address this use case, the best alternative here would be to `skip` that `cleanup` task with a reason. Thoughts?

The notification use case we have been discussing does not apply to this feature in practice because when a task fails, task result of that failed task is not available to the consumer. And to access the execution status of build task, we have another feature [coming soon](https://github.com/tektoncd/pipeline/pull/3390).

@chhsia0 what is the expectation with notification use case in terms of communicating the failure logs?

sbwsg commented:
Hm, resource cleanup and notifications / signalling are the two use cases I'm familiar with.

jerop commented:
@pritidesai i've also seen cleanup and notifications use cases only, maybe we can do a survey (similar to what @sbwsg did for credentials [here](https://groups.google.com/g/tekton-dev/c/n_-NSPRUcm0/m/-IRKv0ZmBgAJ)) to ask users about their use cases for finally tasks (with results and/or when expressions)

vincent-pli commented:
@pritidesai 
I think i can provide another use case, but not definitely.

It's about `Global exception handler` (global in one `pipeline`).
Our user want to handle the exception raised in `tasks` in a `pipeline`, the `task` will not failed when exception raised, we plan to emit the exception to `result`. and in the `final` task, we want a summary of the running situation to do some next job. the reason adopt `final` task rather than normal task stand on the end is because there maybe some unexpected error will make the `task` failed, `final` one will always run. 

So we expect the `result` could be inject to `final` task,  is that a case?

pritidesai commented:
thanks @vincent-pli yes I think this use case fits well with the requirements i.e. as long as the `exception` is communicated through `results` and the `task` producing `exception` does not fail, `finally` task will always run.

The [proposal](https://docs.google.com/document/d/1tV1LgPOINnmlDV-oSNdLB39IlLcQRGaYAxYZjVwVWcs/edit?ts=5fa3127a#heading=h.h3uuz4svzvgs) is almost final now and I am in process of updating [TEP](https://github.com/tektoncd/community/blob/master/teps/0004-task-results-in-final-tasks.md) followed by this PR 🏃‍♀️ 

Thank you everyone for your patience 🙏

sbwsg commented:
/test check-pr-has-kind-label

pritidesai commented:
I have again updated this PR based on the discussion in the [TEP](https://github.com/tektoncd/community/pull/254#issuecomment-728332142), PTAL 🙏

pritidesai commented:
/test pull-tekton-pipeline-integration-tests

pritidesai commented:
This is ready for review, PTAL 🙏

vdemeester reviewed (approved):
/meow

pritidesai commented:
thanks @sbwsg  and @vdemeester for approvals 🎉 

looking for one `lgtm`, @afrittoli, @bobcatfish, @jerop, if you have bandwidth? 🙏

bobcatfish commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -3999,6 +3999,138 @@ func TestReconcilePipeline_TaskSpecMetadata(t *testing.T) {
+	}
+
+	finalTaskSkipped := false
+	for _, s := range reconciledRun.Status.SkippedTasks {
```
would it make sense to also make sure the overall status of the reconciled run is failure in this case? unless that's something you want to cover in an e2e test instead

pritidesai replied:
Reconciled run does not result in failure here, controller logs an error and continues creating taskRun for other final tasks. I have covered the overall pipelineRun status in `e2e` test.

pritidesai replied:
I have added two additional final tasks in `e2e` test [consuming](https://github.com/tektoncd/pipeline/pull/3242/files#diff-7815ee756d1326fe39c09f4deb5c7c36aa4a8f14b4319d5bf8ac73a1929a3d10R94) task result from a failed task and a skipped task (due to when expression). Overall status of this test `pipelineRun` is failure as it has at least one `dag` failed. Let me know if you would like to see one more test pipeline with just dag with when expression and a final task consuming result from that dag which would result in overall pipelineRun status to `completed`.

bobcatfish commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -3999,6 +3999,138 @@ func TestReconcilePipeline_TaskSpecMetadata(t *testing.T) {
+		if s.Name == "final-task-2" {
+			finalTaskSkipped = true
+		}
+	}
```
should we verify that this is also the ONLY task that is skipped?

pritidesai replied:
yup, good point 👍 done

bobcatfish commented on pkg/reconciler/pipeline/dag/dag.go:
```diff
@@ -66,16 +66,29 @@ func (g *Graph) addPipelineTask(t Task) (*Node, error) {
-// Build returns a valid pipeline Graph. Returns error if the pipeline is invalid
-func Build(tasks Tasks) (*Graph, error) {
+// BuildWithoutLinks builds a Pipeline Graph with the specified tasks as nodes (without any links).
+// Returns error if the pipeline is invalid
```
I don't understand what BuildWithoutLinks is doing, could you add more detail, e.g. explain what "links" means in this context? I think it means that the graph is created without any edges, in which case it's not a graph (i mean technically it is but not in a useful way) - is the idea to get a list of nodes? would it make sense to create a different data structure or maybe just return a list of []Node?

pritidesai replied:
`BuildWithoutLinks` is creating graph without linking any dependencies. 

Task results introduces dependency between two tasks i.e. consumer task is dependent on producer task and from the `dag` perspective, both tasks are part of the same graph. But in case of `finally` consuming results from `dag`, both are part of a separate sections. And as per the `finally` contract, it can not depend on any `dag` task.

You are right practically its just a list of `Node`s, all finally tasks being root nodes of a graph.

We could convert it into a list instead of graph. This was originally designed to be a graph to support finally tasks being dependent on (order and resource dependency) other finally tasks.

I have renamed `BuildWithoutLinks` to `BuildWithoutDeps` if its easier to understand.

pritidesai replied:
Another alternative is to have a build function taking both `dag` and `finally` and return two graphs:

```
func Build(dag Tasks, finally Tasks) (*Graph, *Graph, error) {
```

Build graph as is for `dag` tasks but while building graph for `finally`, avoid any dependencies if its a `dag` task.

Or

Build graph with all the `tasks` but avoid adding dependencies if it exists in a list of `noDeps`:

```
func Build(tasks Tasks, noDeps Tasks) (*Graph, error) {
```

Call it like:

```
d := Build(dagTasks, []PipelineTasks)
dfinally := Build(finallyTasks, dagTasks)
```

bobcatfish replied:
Thanks for explaining!

I think it's important to use the right data structure for whatever we're doing - it makes the code much clearer and easier to maintain in the long run.

In this case, if a list of nodes is what we need, do we need to involve a graph object at all? Can we create a function that is not related to Graph at all and produces a list of Nodes from a list of Tasks?

e.g. 

```
func GetNodes(tasks Tasks) []Node
```

Alternatively: maybe the problem is that determining the deps shouldnt be part of building the graph? what if `deps` was passed into the build function?

pritidesai replied:
> I think it's important to use the right data structure for whatever we're doing - it makes the code much clearer and easier to maintain in the long run.

yup, building `Graph` for `finally` section is essential to support `runAfter` and `workspaces` in future.

pritidesai replied:
> Alternatively: maybe the problem is that determining the deps shouldnt be part of building the graph? what if deps was passed into the build function?

yup, working on this alternative.

pritidesai replied:
Taking `deps` out of `build` does simplify `build`, see changes [here](https://github.com/tektoncd/pipeline/compare/master...pritidesai:build-deps?expand=1) 

And, irrespective of `finally` section being built as a `Graph` or a list of `Node[]`, it has to fetch `deps` to validate validate result reference belongs to a valid `dag` task. 

PTAL 🙏

bobcatfish replied:
Nice! I wonder if it would help if there was a function you could call to get all the deps as a map?

e.g. something like

```

type PipelineTasks []PipelineTask
func (tasks PipelineTasks) Deps() map[string][] {
        deps := map[string][]string{}
	for _, pt := range tasks {
		deps[pt.HashKey()] = pt.Deps()
	}
        return deps
}
```

then the original line:

```
if _, err := dag.Build(PipelineTaskList(tasks)); err != nil {
```

would just become something like :

```
if _, err := dag.Build(PipelineTaskList(tasks),  tasks.Deps()); err != nil {
```

bobcatfish replied:
(oh! `PipelineTaskList` already IS the type I defined above as `PipelineTasks` :D)

pritidesai replied:
Great suggestion, it looks much simpler now, PR # https://github.com/tektoncd/pipeline/pull/3583

pritidesai replied:
> validate the result reference belonging to a valid dag task which is missing from the changes in this PR 😢

I have added validation [here](https://github.com/tektoncd/pipeline/pull/3242/files#diff-78ff878240189b20e0ef82f2912c93c0b3c7b445865e04fa9b98aab851be6130R343).

bobcatfish reviewed (commented):
Hey @pritidesai ! I've looked at the end to end and reconciler tests and I started to look at the code (just BuildWithoutLinks so far).

Here is my feedback so far:

* what do you think about adding some use of results to the existing e2e test(s) as well? https://github.com/tektoncd/pipeline/blob/master/test/pipelinefinally_test.go
* I'm also thinking that based on some of the bugs we ran into with when expressions, it might make sense to create a more complex example, e.g. maybe including a when expression somewhere in the finally example as well (any other features we might want to include?)
* Similar to the previous point, it seems like most of the tests are using a case where a Task producing a result fails - can we include some where it is skipped as well? In this case, the overall PipelineRun should still succeed (if im understanding correctly!) but the skipped task and the final task that needs the result should both be skipped

pritidesai commented on docs/pipelines.md:
````diff
@@ -759,6 +759,36 @@ spec:
+          value: $(tasks.count-comments-before.results.count)
+        - name: after-count
+          value: $(tasks.count-comments-after.results.count)
+```
````
I have replaced an example with more common task `git-clone` so that its easier to follow.

pritidesai commented on docs/pipelines.md:
```diff
@@ -759,6 +759,36 @@ spec:
+    - name: add-comment
+      taskRef:
+        Name: add-comment
+    - name: count-comments-after
```
I have replaced an example with more common task `git-clone` so that its easier to follow.

pritidesai commented:
This is rebased and ready for review 🙏

pritidesai commented:
/test check-pr-has-kind-label

pritidesai commented:
> pinging this again
Gave it a one more push, have dropped using test builder in `pipelinerun_test.go`.

bobcatfish commented on docs/pipelines.md:
```diff
@@ -760,6 +760,33 @@ spec:
+`clone-app-repo` failed or skipped with [when expression](#guard-task-execution-using-whenexpressions) resulting in
+uninitialized task result `commit`, the finally Task `discover-git-commit` will be included in the list of
+`skippedTasks`. The controller logs the validation failure `unable to find result referenced by param "commit" in
+"discover-git-commit"` and continues executing rest of the final tasks. The pipeline exits with `completion` instead of
```
note: i really dont think we should include anything about what the controller will likely log in these docs. if we start documenting that kind of thing, then users might come to rely on it.

imo the docs should cover the expected functionality that would be more or less protected by our backwards compatibility policy - if we wanted to change what the controller logs, i dont think we should need to change our docs.

i also think it's important that we distinguish between:
- how the features are implemented
- what the features do

users need to know about what the features do, and as much as possible shouldn't have to think about how the features are implemented (which also leaves us maintainers free to change the implementation around without impacting users)

pritidesai replied:
I can remove this log from here for now but without this log, if I am running into a finally task being part of `skippedTasks`, I have no way to look for `why` from doc perspective (without looking at the code).

pritidesai replied:
deleted log from doc.

bobcatfish commented on test/yamls/v1beta1/pipelineruns/pipelinerun-with-final-tasks.yaml:
```diff
@@ -164,6 +164,18 @@ spec:
+          - name: display-commit
+            image: alpine
+            script: |
+              echo $(params.commit)
```
im confused, why does a yaml file with this name exist in 2 different places?

pritidesai replied:
Please refer to https://github.com/tektoncd/pipeline/pull/3564#issuecomment-736252906

bobcatfish commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -4350,6 +4350,188 @@ func TestReconcilePipeline_TaskSpecMetadata(t *testing.T) {
+	if d := cmp.Diff(expectedTaskRun, actualTaskRun, ignoreResourceVersion); d != "" {
+		t.Errorf("expected to see TaskRun %v created. Diff %s", expectedTaskRunName, diff.PrintWantGot(d))
+	}
+
```
continuing on this conversation:

> me: would it make sense to also make sure the overall status of the reconciled run is failure in this case? unless that's something you want to cover in an e2e test instead

> @pritidesai: Reconciled run does not result in failure here, controller logs an error and continues creating taskRun for other final tasks. I have covered the overall pipelineRun status in e2e test.

I think I almost understand, but I think technically the situation is that the controller is waiting for the finally tasks to complete before updating the status to failed, is that right?

> controller logs an error and continues creating taskRun for other final tasks

this reconcile should result in all finally tasks "running" (since they run in parallel) and then the it would the the last reconcile after the finally tasks have finished that would set the overall status to failed, is that right?

pritidesai replied:
> I think I almost understand, but I think technically the situation is that the controller is waiting for the finally tasks to complete before updating the status to failed, is that right?

Yes the controller is waiting for the finally tasks to complete before update the status. The status condition need not be necessarily failure. If finally task has a reference to a skipped dag task, the controller sets the status to `completed` after all finally tasks are finished. The controller sets the status to `failed` if any of the finally or a dag task fails. Let me know if this is not clear.

bobcatfish commented on test/pipelinefinally_test.go:
```diff
@@ -55,25 +58,69 @@ func TestPipelineLevelFinally_OneDAGTaskFailed_Failure(t *testing.T) {
+
+	f := map[string]pipelineTask{
+		"finaltask1": {
+			TaskName: successTask.Name,
```
would it make sense to have a final task also successfully consuming a result, for completeness?

pritidesai replied:
haha I had added it and dropped it, sure will add it back.

pritidesai replied:
done

bobcatfish commented on test/pipelinefinally_test.go:
```diff
@@ -110,11 +163,15 @@ func TestPipelineLevelFinally_OneDAGTaskFailed_Failure(t *testing.T) {
 			if !isSkipped(t, n, taskrunItem.Status.Conditions) {
 				t.Fatalf("dag task %s should have skipped due to condition failure", n)
 			}
+		case n == "dagtask4":
```
nitpick but maybe it would make sense to also assert on the length of taskrunList? apologies since this test was already like this, but it seems like it would currently pass if extra taskruns were returned

pritidesai replied:
sure I will add it explicitly, it is covered by the default:

https://github.com/tektoncd/pipeline/blob/705e97146fea0e6a7f55e3b06f46e317a9c12659/test/pipelinefinally_test.go#L118-L119

`default` would raise error if it finds any missing or extra taskRuns, I will add explicit count assertion.

pritidesai replied:
will update this error message as well since its not clear.

pritidesai replied:
https://github.com/tektoncd/pipeline/pull/3242/files#diff-7815ee756d1326fe39c09f4deb5c7c36aa4a8f14b4319d5bf8ac73a1929a3d10R168

bobcatfish commented on test/pipelinefinally_test.go:
```diff
@@ -123,6 +180,26 @@ func TestPipelineLevelFinally_OneDAGTaskFailed_Failure(t *testing.T) {
+		t.Fatalf("Expected four skipped tasks, dag task with condition failure dagtask3, dag task with when expression,"+
+			"two final tasks with missing result reference finaltaskconsumingdagtask1 and finaltaskconsumingdagtask4 in SkippedTasks."+
+			" Diff: %s", diff.PrintWantGot(d))
+	}
```
thanks for updating this test!!

bobcatfish commented on test/pipelinefinally_test.go:
```diff
@@ -123,6 +180,26 @@ func TestPipelineLevelFinally_OneDAGTaskFailed_Failure(t *testing.T) {
+	}
 }
 
 func TestPipelineLevelFinally_OneFinalTaskFailed_Failure(t *testing.T) {
```
do we need to update more than one end to end test? i feel like the previous one was pretty comprehensive, especially if it also has a finally task successfully execute while consuming a result

pritidesai replied:
nope, I think updating one is enough, I will add a finally task with a successful task result consumption in the first test itself.

pritidesai replied:
done

pritidesai replied:
verifying the param value in a finally task (referring to result of a dag task) https://github.com/tektoncd/pipeline/pull/3242/files#diff-7815ee756d1326fe39c09f4deb5c7c36aa4a8f14b4319d5bf8ac73a1929a3d10R210

bobcatfish commented on pkg/reconciler/pipelinerun/pipelinerun.go:
```diff
@@ -589,7 +589,19 @@ func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.Pip
 	pipelineRunFacts.ResetSkippedCache()
 
 	// GetFinalTasks only returns tasks when a DAG is complete
-	nextRprts = append(nextRprts, pipelineRunFacts.GetFinalTasks()...)
```
this is outside of the scope of this PR but i wonder if this function (`runNextSchedulableTask`) has too much responsibility. it seems to be in charge of:
- deciding what to execute next
- resolving references
- actually executing

those feel like at least 3 distinct responsibilities. i think if we get a chance it would be great to detangle this a bit.

bobcatfish replied:
not to mention that it doesnt have unit tests :(

pritidesai replied:
😭 it does

I will work on refactoring it after this is merged, created issue #3610

bobcatfish commented on pkg/reconciler/pipelinerun/pipelinerun.go:
```diff
@@ -589,7 +589,19 @@ func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.Pip
+	for _, rprt := range finallyRprts {
+		resolvedResultRefs, err := resources.ResolveResultRef(pipelineRunFacts.State, rprt)
+		if err != nil {
+			logger.Warnf("Final task %q is not executed as it failed to resolve task params for %q with error %v", rprt.PipelineTask.Name, pr.Name, err)
```
should this be a warning or an error? we only need to apply results when we're ready to run final tasks and we're only ready once all of the non-final tasks have finished.

bobcatfish replied:
im thinking that what is happening here is that we are logging this warning in the case where a finally task needs a result from a task that was skipped.

1. i dont think that should be a warning. logs that are warn, error or fatal should be indicating to the operator of the system that some kind of action needs to be taken. i.e. the _operator_ not the _user_. the user gets informed of actions they need to take via the status (e.g. error messages). the operator is the person admin-ing the tekton installation and needs to know if there is something wrong with the setup. so maybe info here? (https://dzone.com/articles/logging-levels-what-they-are-and-how-they-help-you seems like an okay article on log levels)
2. it would be great if one day we could get to a point where the decision of whether or not to skip was only done in one place - is there any way to do that in this PR? i.e. before we try to resolve the references? (i.e. why are we trying to resolve references on something that needs to be skipped) it seems like we should be able to look at a task's dependencies and see that it cant be run?

pritidesai replied:
> should this be a warning or an error?

I think it should be warning, its indicating that the finally task was skipped for some reason. Skipping a dag does not warn users but here users are warned when a finally task is skipped. Finally being finally task must be executed and warn if they are not for some reason.

But based on your description, there is no action needs to be taken if a finally task is skipped in case of a cleanup/notification use case. may be convert it to info then? I am not convinced but at the same time I dont see a use case.


> im thinking that what is happening here is that we are logging this warning in the case where a finally task needs a result from a task that was skipped.

yup skipped or failed.

> we only need to apply results when we're ready to run final tasks and we're only ready once all of the non-final tasks have finished.

Until we are ready to run finally tasks, `finallyRprts` is empty.

> it would be great if one day we could get to a point where the decision of whether or not to skip was only done in one place - is there any way to do that in this PR?

The only way we could do it to maintain a list of [skippedTasks](https://github.com/tektoncd/pipeline/compare/master...pritidesai:pipelinerunstatus#diff-188e8bbe2f5c7cb0b9de03be02a957537e46bb71add977e4992533375ef08998R351) in `pipelineRunFacts`. Once we determine that a finally task is being skipped, add it into the list. And check this list while updating the status condition and status.SkippedTasks. The same logic would work for dag tasks as well. We could merge this as is and refactor after this is merged.

(@afrittoli's attempt on simplifying [skipped](https://github.com/tektoncd/pipeline/commit/72455ba0a81c9d806b26809842282c045ca43d41#diff-2d6877ada24d6e8ae6d232d24799f21c520e590e5b432f9adba5e2e62eaa40b7R53) tasks)

> why are we trying to resolve references on something that needs to be skipped

At this point, we are discovering task result resolution failure and therefore not executing finally task. Finally task consuming task results from a dag task but not dependent on a dag task. Both belong to separate sections, no relationships established from a finally to dag.

pritidesai replied:
This could be further simplified for both finally and dag, instead of trying to resolve task results, just check if the referenced task was failed or skipped before trying to resolve results.

bobcatfish commented on pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:
```diff
@@ -204,6 +204,19 @@ func (t *ResolvedPipelineRunTask) parentTasksSkip(facts *PipelineRunFacts) bool
 }
 
+// IsFinallySkipped returns true if a finally task is not executed and skipped due to task result validation failure
+func (t *ResolvedPipelineRunTask) IsFinallySkipped(facts *PipelineRunFacts) bool {
```
im surprised we need special "is skipped" logic for a finally task - wouldnt the logic for whether or not a task is skipped be the same as for non-finally tasks? (i.e. it depends on a task that failed or was skipped?)

bobcatfish reviewed (commented):
I was able to get into the code this time :D very excited to see this progressing!

pritidesai commented:
Hey @bobcatfish I think I have responded to most of the comments. I have tried to explain both of these:

(1) Need for a separate `isFinallySkipped` function and possible refactoring. Can be avoided by introducing a list of `skippedTasks` as part of `pipelineRunFacts`. 
(2) Checking if a finally task must be skipped in one single i.e. instead of resolving a task result, check the reference task status, if failed or skipped, skip finally

Let me know if you would want to see any more changes in this PR including any refactoring. I will be happy to change this PR or create separate PR if possible.

pritidesai commented:
one more ping 🛎
🙏🏻 Would like to get this in if possible 🙏🏻

pritidesai commented:
> /lgtm cancel
> /hold

Hey @vdemeester we discussed this in the planning meeting this morning. I think we put this PR on hold so that it can be redesigned (skip instead of creating an object reference for a finally task referring to missing task results). Can we cancel the hold  since the design is updated?

We have moved this PR to next milestone 0.21 for now.

vdemeester commented:
> > /lgtm cancel
> > /hold
> 
> Hey @vdemeester we discussed this in the planning meeting this morning. I think we put this PR on hold so that it can be redesigned (skip instead of creating an object reference for a finally task referring to missing task results). Can we cancel the hold since the design is updated?
> 
> We have moved this PR to next milestone 0.21 for now.

👍🏼 

/hold cancel

vdemeester commented:
@pritidesai needs a rebase 🙏🏼

pritidesai commented:
> @pritidesai needs a rebase 🙏🏼

sorry @vdemeester for delay, done 👍

vdemeester commented:
/lgtm
